[win] Replace the .seh_startchained and .sehendchained instructions with .seh_splitchained

dpaoliello opened the pull request:
LLVM's existing support for chaining x64 Windows SEH frame infos is broken. Windows requires that each frame info (including the parent) is adjacent, but not overlapping, and the whole function is covered (i.e., each instruction maps to exactly one frame info). Therefore, having start/end chaining pseudo instructions doesn't make any sense, as every "end chain" pseudo would need a start pseudo immediately after it or be at the end of function.

This change switches having a "split chain" pseudo instruction that ends the current frame and starts a new chained frame.

Added a release note about the replacement - to my knowledge there is no one actually using frame info chaining, so it is highly unlikely that it will break any code.

Split off from #159206

dpaoliello committed:
[win] Replace the .seh_startchained and .sehendchained instructions with .seh_splitchained

efriedma-quic reviewed (commented):
Please add a testcase showing how interactions like the following work.

```
.seh_pushreg %rbx
.seh_endprologue
.seh_splitchained
.seh_pushreg %rsi
.seh_endprologue
```

--------

Please add a testcase showing the interaction between unwind-v2 and seh_splitchained.  (I think you already had a testcase, but just didn't add it here for some reason?)

dpaoliello commented:
> Please add a testcase showing how interactions like the following work.
> 
> ```
> .seh_pushreg %rbx
> .seh_endprologue
> .seh_splitchained
> .seh_pushreg %rsi
> .seh_endprologue
> ```

Done, added to `seh.s`.

> Please add a testcase showing the interaction between unwind-v2 and seh_splitchained. (I think you already had a testcase, but just didn't add it here for some reason?)

I wasn't sure how much of the unwindv2 stuff you wanted in this PR, but it makes sense that using `.seh_splitchained` shouldn't break unwindv2, so I've added the changes to support chaining and the associated tests.

efriedma-quic commented on llvm/lib/MC/MCStreamer.cpp:
```diff
@@ -994,7 +994,8 @@ void MCStreamer::emitWinCFIBeginEpilogue(SMLoc Loc) {
 
-  if (!CurFrame->PrologEnd)
+  // Chained unwinds aren't guaranteed to have a prolog.
+  if (!CurFrame->PrologEnd && !CurFrame->ChainedParent)
```
Do we need to do something to prevent someone from writing:

```
    .seh_endprologue
    .seh_splitchained
    pushq %rsi
    .seh_pushreg %rsi
    .seh_startepilogue
```

dpaoliello commented:
Ping @efriedma-quic any other concerns?

efriedma-quic reviewed (commented):
Do we want to diagnose something like following?

```
    .seh_pushreg %rbx
    .seh_endprologue
    .seh_splitchained
    pushq %rsi
    .seh_pushreg %rsi
    .seh_splitchained
    .seh_endprologue
    .seh_startepilogue
```

The rules might be easier to understand if we just make .seh_endprologue mandatory after .seh_splitchained.

dpaoliello commented:
> The rules might be easier to understand if we just make .seh_endprologue mandatory after .seh_splitchained.

I was going to push back on this, but the more I thought about it, the more I like it. It definitely simplifies the logic in the parser, and it makes the story for hand-written code simpler to explain: "As soon as an unwind frame is started (either at the beginning of a func or in a chained frame), you are implicitly in the prolog and must explicitly end the prolog whether you have any instructions or not".

efriedma-quic reviewed (approved):
LGTM
